rpc/kate-rpc/src/lib.rs

```diff
@@ -118,6 +126,88 @@ where
 	SDFilter: submitted_data::Filter<CallOf<Block>>,
 	<<Block as BlockT>::Extrinsic as Extrinsic>::Call: Clone,
 {
+	fn query_app_data(
+		&self,
+		app_id: AppId,
+		at: Option<HashOf<Block>>,
+	) -> Result<Vec<Option<Vec<u8>>>> {
+		let block_id = self.block_id(at);
+
+		let signed_block = self
+			.client
+			.block(&block_id)
+			.map_err(|e| internal_err!("Invalid block number: {:?}", e))?
+			.ok_or_else(|| internal_err!("Missing block {}", block_id))?;
+		let block_hash = signed_block.block.header().hash();
+
+		let mut block_ext_cache = self
+			.block_ext_cache
+			.write()
+			.map_err(|_| internal_err!("Block cache lock is poisoned .qed"))?;
+
+		let block_length: BlockLength = self
+			.client
+			.runtime_api()
+			.get_block_length(&block_id)
+			.map_err(|e| internal_err!("Block Length cannot be fetched: {:?}", e))?;
+
+		let rows: usize = block_length.rows as usize;
+		let cols: usize = block_length.cols as usize;
+		let chunk_size: usize = block_length.chunk_size() as usize;
+
+		if !block_ext_cache.contains(&block_hash) {
+			// build block data extension and cache it
+			let xts_by_id: Vec<AppExtrinsic> = signed_block
+				.block
+				.extrinsics()
+				.iter()
+				.map(|e| AppExtrinsic {
+					app_id: e.app_id(),
+					data: e.encode(),
+				})
+				.collect();
+
+			// Use Babe's VRF
+			let seed: [u8; 32] = self
+				.client
+				.runtime_api()
+				.get_babe_vrf(&block_id)
+				.map_err(|e| internal_err!("Babe VRF not found for block {}: {:?}", block_id, e))?;
+
+			let (_, block, block_dims) =
+				kate::com::flatten_and_pad_block(rows, cols, chunk_size, &xts_by_id, seed)
+					.map_err(|e| internal_err!("Flatten and pad block failed: {:?}", e))?;
+
+			let data = kate::com::par_extend_data_matrix(block_dims, &block)
+				.map_err(|e| internal_err!("Matrix cannot be extended: {:?}", e))?;
+
+			block_ext_cache.put(block_hash, (data, block_dims));
+		}
+
+		let (ext_data, block_dims) = block_ext_cache
+			.get(&block_hash)
+			.ok_or_else(|| internal_err!("Block hash {} cannot be fetched", block_hash))?;
+
+		let HeaderExtension::V1(header) = signed_block.block.header().extension();
+		let DataLookup { index, size } = &header.app_lookup;
+
+		let app_data_index = AppDataIndex {
```

Comment:
Types `AppDataIndex` and `DataLookupIndexItem` are the same. Could we merge them?

Reply:
I guess as soon as we decouple avail core from substrate we can reuse data. At this point, merging it would require to depend on sp-runtime on LC.